src/sim/TrackInterface.hh

```diff
@@ -18,54 +18,109 @@
 {
-    template<template<Ownership, MemSpace> class S>
-    using DeviceCRef = S<Ownership::const_reference, MemSpace::device>;
-
-    DeviceCRef<GeoParamsData>      geo;
-    DeviceCRef<MaterialParamsData> material;
-    DeviceCRef<ParticleParamsData> particle;
+    GeoParamsData<W, M>      geometry;
+    MaterialParamsData<W, M> materials;
+    ParticleParamsData<W, M> particles;
+    RngParamsData<W, M>      rng;
 
-    //! Whether the data are assigned
+    //! True if all params are assigned
     explicit CELER_FUNCTION operator bool() const
     {
-        return geo && material && particle;
+        return geometry && materials && particles;
+    }
+
+    //! Assign from another set of data
+    template<Ownership W2, MemSpace M2>
+    ParamsData& operator=(const ParamsData<W2, M2>& other)
+    {
+        CELER_EXPECT(other);
+        geometry  = other.geometry;
+        materials = other.materials;
+        particles = other.particles;
+        rng       = other.rng;
+        return *this;
     }
 };
 
-//---------------------------------------------------------------------------//
-// STATE
 //---------------------------------------------------------------------------//
 /*!
  * Thread-local state data.
  */
-struct StatePointers
+template<Ownership W, MemSpace M>
+struct StateData
 {
-    template<template<Ownership, MemSpace> class S>
-    using DeviceRef = S<Ownership::reference, MemSpace::device>;
+    template<class T>
+    using Items = StateCollection<T, W, M>;
+
+    GeoStateData<W, M>      geometry;
+    ParticleStateData<W, M> particles;
+    RngStateData<W, M>      rng;
+    SimStateData<W, M>      sim;
 
-    DeviceRef<ParticleStateData> particle;
-    DeviceRef<RngStateData>      rng;
-    DeviceRef<GeoStateData>      geo;
-    DeviceRef<SimStateData>      sim;
+    // Raw data
+    Items<celeritas::Interaction> interactions;
 
-    Span<Interaction> interactions;
+    //! Number of tracks
+    CELER_FUNCTION celeritas::size_type size() const
+    {
+        return particles.size();
+    }
 
     //! Whether the data are assigned
     explicit CELER_FUNCTION operator bool() const
     {
-        return particle && geo && sim && rng && !interactions.empty();
+        return geometry && particles && rng && sim && !interactions.empty();
     }
 
-    //! Number of tracks
-    CELER_FUNCTION size_type size() const { return particle.size(); }
+    //! Assign from another set of data
+    template<Ownership W2, MemSpace M2>
+    StateData& operator=(StateData<W2, M2>& other)
+    {
+        CELER_EXPECT(other);
+        geometry     = other.geometry;
+        particles    = other.particles;
+        rng          = other.rng;
+        sim          = other.sim;
+        interactions = other.interactions;
+        return *this;
+    }
 };
 
+using ParamsDeviceRef
+    = ParamsData<Ownership::const_reference, MemSpace::device>;
+using StateDeviceRef = StateData<Ownership::reference, MemSpace::device>;
+
+#ifndef __CUDA_ARCH__
```

Comment:
As of #226 we no longer have to bother with the CUDA_ARCH complier exclusion, to simplify things

Reply:
Forgot about that, thanks!